plugins/module_utils/identity/bitwarden.py

```diff
@@ -0,0 +1,321 @@
+        self._folder = Folder(id_=value['folderId'])
+
+
+class Folder(_Element):
+    '''A folder in the password database.'''
+
+    organization = property(lambda self: self._organization)
+
+    def __init__(self, name=None, id_=None, organization=None, dict_=None):
+        if [name is None, id_ is None, dict_ is None].count(False) > 1:
+            raise AssertionError('Cannot specify folder in multiple ways')
+
+        self._name = name
+        self._id = id_
+        self._organization = organization
+
+        if dict_:
+            self.dict = dict_
+
+    def _by_attribute(self, attribute, value):
+        '''Return folder dict where the specified attribute
+        has the specified value.
+        '''
+        matches = [folder for folder in self.organization.folders if folder.dict[attribute] == value]
+
+        if not matches:
+            raise KeyError("Folder not found")
+        if len(matches) > 1:
+            raise AssertionError("Folder not unique")
+
+        return matches[0]
+
+    @property
+    def items(self):
+        query = ['list', 'items']
+
+        # Only restrict by folder if specified.
+        if self.name is not None:
+            # bw treats 'null' as items not belonging to any folder.
+            query.extend(['--folderid', self.id or 'null'])
+
+        # Filter by organization since `bw` ORs filters together.
+        return [Item(dict_=d) for d in Client.run_json(query) if self.organization.matches(d)]
+
+    def get_item(self, name=None, id_=None):
+        return Item(name, id_, self, self.organization)
+
+    def matches(self, folder):
+        '''Return whether folder matches a pattern indicated by self.
+
+        If self.name is None, any folder matches.
+        If self.name is '' or 'No Folder', only the null folder is matched.
+        All other folders are matched by id.
+        '''
+
+        if self.name is None:
+            return True
+
+        # No Folder is a special folder name in bw.
+        if self.name in ('', 'No Folder'):
+            return folder['folderId'] is None
+
+        return self.id == folder['folderId']
+
+
+class Organization(_Element):
+    '''An organization in the password database.'''
+
+    def __init__(self, name=None, id_=None, dict_=None):
+        if [name is None, id_ is None, dict_ is None].count(False) > 1:
+            raise AssertionError('Cannot specify organization in multiple ways')
+
+        self._name = name
+        self._id = id_
+
+        if dict_:
+            self.dict = dict_
+
+    def _by_attribute(self, attribute, value):
+        '''Return organization dict where the specified attribute
+        has the specified value.
+        '''
+
+        matches = [org for org in organizations() if org.dict[attribute] == value]
+
+        if not matches:
+            raise KeyError("Organization not found")
+        if len(matches) > 1:
+            raise AssertionError("Organization not unique")
+
+        return matches[0]
```

Comment:
Ah. Good catch! In an earlier incarnation of this code (back before I extracted the subprocess spawning code), this did actually use `self`.

Thoughts on turning this into a `@staticmethod`? I'm not sure I understand your second concern, but perhaps that addresses it? I originally opted to keep the method as part of the `Organization` class for consistency with the API for `Folder` and `Item`.

I can also group the `organizations()` function as a static method in `Organization`. The thing I was trying to avoid for legibility before (metaclass programming) is no longer an issue. Any thoughts on that?

Reply:
Concern 2 refers to the fact that you are breaking the encapsulation of that element, by accessing something external to the class. Which is pretty much the same reason why we should avoid global variables in general.

Concern 3 refers to the fact that your class represents (or should represent) ONE Organization. Or, if you want it to represent both ONE Org *and* the complete LIST of ORGs, then be explicit about it. In that case, `organizations()` should definitely be inside the class. But if you are going to do that for one class, you might as well make it generic (maybe `@classmethod`) so that every element class can manage its own list.

Right now it is neither one thing nor the other.

Reply:
Ah, I think I see. Thank you for clarifying!

I've found the behaviour of the CLI (and the web frontend) to be very unpredictable in terms of folders -- folders sometimes have one UUID but appear in multiple organizations and sometimes they appear in only one organization. I tried (not very hard) to figure out the rules for this. Folder and items also appear to have some weird interaction. Only organization IDs appeared well-behaved to me. As such, I don't think I can implement anything that will behave as expected for folders or items. And it would certainly be different enough for organizations/folders/items that it would require an entirely separate implementation for each. If there are no objections, I'm only going to `@staticmethod` the function.